Add requireEmailVerification argument to $requireSignIn() firebase auth router helper

ericmorgan1 opened the pull request:
<!--

Thank you for contributing to the Firebase community! Please fill out the pull request form below
and make note of the following:

Run the linter and test suite
==============================
Make sure your changes pass our linter and the tests all pass on your local machine. We've hooked
up this repo with continuous integration to double check those things for you.

Add tests (if applicable)
==============================
Most non-trivial changes should include some extra test coverage. If you aren't sure how to add
tests, feel free to submit regardless and ask us for some advice.

Sign our CLA
==============================
Please sign our Contributor License Agreement (https://cla.developers.google.com/about/google-individual)
before sending PRs. We cannot accept code without this.

-->


### Description

This adds an argument (requireEmailVerification) to the $requireSignIn router helper in the angularfire auth service.
When set to true, this will fail if the firebase auth "emailVerified" property is false.
This is useful if you do not want users to access your app until they have verified their email address.

### Code sample

```javascript
_routerMethodOnAuthPromise: function(rejectIfAuthDataIsNull, rejectIfEmailNotVerified) {
      var self = this;

      // wait for the initial auth state to resolve; on page load we have to request auth state
      // asynchronously so we don't want to resolve router methods or flash the wrong state
      return this._initialAuthResolver.then(function() {
        // auth state may change in the future so rather than depend on the initially resolved state
        // we also check the auth data (synchronously) if a new promise is requested, ensuring we resolve
        // to the current auth state and not a stale/initial state
        var authData = self.getAuth(), res = null;
        if (rejectIfAuthDataIsNull && authData === null) {
          res = self._q.reject("AUTH_REQUIRED");
        }
        else if (rejectIfEmailNotVerified && !authData.emailVerified) {
            res = self._q.reject("EMAIL_NOT_VERIFIED");
        }
        else {
          res = self._q.when(authData);
        }
        return res;
      });
    },
```

katowulf commented:
I wonder if this would be more appropriate as an argument to $requireSignIn? 

Is there a case where you would require email verification but not sign in? I don't think there is.

ericmorgan1 commented:
Good point! You're right it wouldn't make sense to need email verification but not sign in.
I updated it-- what do you think?

katowulf commented:
Looks good! Could we also get this into the [test units](https://github.com/firebase/angularfire/blob/master/tests/unit/FirebaseAuth.spec.js#L355)?

ericmorgan1 commented:
Sure! :)

Btw, have you seen this error?
I'm getting it when running the tests, but I can look into what's going on as well.

      √ should return true once $loaded() promise is rejected
    $ref
      √ should return Firebase instance it was created with
Chrome 56.0.2924 (Windows 8.1 0.0.0) ERROR
  Uncaught Possibly unhandled rejection: {}
  at .../angularfire/node_modules/angular-mocks/angular-mocks.js:250
Chrome 56.0.2924 (Windows 8.1 0.0.0) ERROR
  Uncaught Possibly unhandled rejection: {}
  at .../angularfire/node_modules/angular-mocks/angular-mocks.js:250

Chrome 56.0.2924 (Windows 8.1 0.0.0): Executed 50 of 274 ERROR (6.772 secs / 6.475 se
cs)

jwngr commented on src/auth/FirebaseAuth.js:
```diff
@@ -200,6 +202,9 @@
           res = self._q.reject("AUTH_REQUIRED");
         }
+        else if (rejectIfEmailNotVerified && !authData.emailVerified) {
+            res = self._q.reject("EMAIL_NOT_VERIFIED");
```
Suggestion: `"EMAIL_VERIFICATION_REQUIRED"`

ericmorgan1 commented:
Thanks for the help!

I added some unit tests and updated the error code as you suggested.

All the Auth tests passed, but I still couldn't get the others working, so I ignored them.

FYI:
I was using node 5.2.0, but updated to 6.9.2, but still got that error.
My Chrome version is 56.0.2924.28 beta (64-bit).
Not very sure how to debug it, but if you have any suggestions, I'll try it out!

jwngr commented:
Looks great! Thanks for your contribution! I'm going to hold off on merging this in just yet since it updates documentation and I want to time it with the next release. I'll cut a new release before I head off for the holidays at the end of this week.

I'll also look into the testing issues. I was able to reproduce them with Node.js 4.4.5 / Chrome 55.0.2883. Thanks for the added setup info you shared.

jwngr commented:
FYI I just fixed the unit test issue with https://github.com/firebase/angularfire/pull/890.

ericmorgan1 commented:
Great! Thank you both for all the help and for looking at this so quickly! :)

jwngr commented:
This has been released as part of [AngularFire `2.2.0`](https://github.com/firebase/angularfire/releases/tag/v2.2.0). Thanks again for your contribution!
